File: 6lab/cpp/Order.cpp
#include "Order.h"
#include <iostream>
#include <iomanip>
#include <cstdlib>
#include <ctime>
#include <fstream>
#include <sstream>
#include <limits>
#include <algorithm>

using namespace std;

#ifdef _MSC_VER
#define _CRT_SECURE_NO_WARNINGS
#endif

// Конструктор по умолчанию
Order::Order() : itemCount(0), countOrder(0), costOrder(0.0) {
    // Инициализация массива указателей
    for (int i = 0; i < MAX_ORDER_ITEMS; i++) {
        orderItems[i] = nullptr;
        quantities[i] = 0;
    }

    numberOrder = generateOrderNumber();
    statusOrder = "Ожидает оплаты";
    dateOrder = getCurrentDate();
    shopOrder = "Музыкальный магазин 'MusicLoverzz'";
    payWay = "Не выбран";

    cout << "Вызван конструктор по умолчанию Order: " << numberOrder << endl;
}

// Конструктор с параметрами
Order::Order(const std::string& clientId, Basket& basket) :
    clientOrder(clientId), itemCount(0), countOrder(0), costOrder(0.0) {

    // Инициализация массива указателей
    for (int i = 0; i < MAX_ORDER_ITEMS; i++) {
        orderItems[i] = nullptr;
        quantities[i] = 0;
    }

    numberOrder = generateOrderNumber();
    statusOrder = "Ожидает оплаты";
    dateOrder = getCurrentDate();
    shopOrder = "Музыкальный магазин 'MusicLoverzz'";
    payWay = "Не выбран";

    CopyProductsFromBasket(basket);

    cout << "Вызван конструктор с параметрами Order: " << numberOrder << endl;
}

// Деструктор
Order::~Order() {
    cout << "Вызван деструктор Order: " << numberOrder << endl;
    clearOrderItems();
}

// Очистка массива товаров
void Order::clearOrderItems() {
    for (int i = 0; i < itemCount; i++) {
        if (orderItems[i] != nullptr) {
            delete orderItems[i];
            orderItems[i] = nullptr;
        }
    }
    itemCount = 0;
}

// Копирование товаров из корзины
void Order::CopyProductsFromBasket(Basket& basket) {
    if (basket.IsEmpty()) {
        return;
    }

    itemCount = basket.getProductCount();
    countOrder = basket.GetBasketCount();
    costOrder = basket.GetBasketCost();

    for (int i = 0; i < itemCount; i++) {
        Product* basketProduct = basket.getProduct(i);
        if (basketProduct != nullptr) {
            orderItems[i] = new Product(*basketProduct);
            quantities[i] = basket.getQuantity(i);
        }
    }
}

// Статический метод для создания заказа
bool Order::MakeOrder(const std::string& clientId, Basket& basket) {
    if (basket.IsEmpty()) {
        cout << "Корзина пуста! Добавьте товары перед оформлением заказа." << endl;
        return false;
    }

    char confirmation;
    cout << "\n=== ПОДТВЕРЖДЕНИЕ ЗАКАЗА ===" << endl;
    basket.DisplayBasket();
    cout << "\nВы точно хотите оформить заказ? (y/n): ";
    cin >> confirmation;

    if (confirmation != 'y' && confirmation != 'Y') {
        cout << "Оформление заказа отменено." << endl;
        return false;
    }

    // Создаем заказ
    Order* newOrder = new Order(clientId, basket);

    if (newOrder->Payment()) {
        newOrder->ChangeStatus("Оплачен");
        cout << "\n=== ЗАКАЗ УСПЕШНО ОФОРМЛЕН ===" << endl;
        newOrder->InfoOrder();
        newOrder->SaveToFile();
        basket.ClearBasket();
        delete newOrder;
        return true;
    }
    else {
        newOrder->ChangeStatus("Ошибка оплаты");
        cout << "Ошибка при оформлении заказа!" << endl;
        delete newOrder;
        return false;
    }
}

// Сохранение заказа в файл
void Order::SaveToFile() {
    ofstream file("orders.txt", ios::app);
    if (file.is_open()) {
        file << numberOrder << "|" << clientOrder << "|" << statusOrder << "|" << costOrder;
        for (int i = 0; i < itemCount; i++) {
            if (orderItems[i] != nullptr) {
                file << "|" << orderItems[i]->getArticle() << "|" << quantities[i];
            }
        }
        file << endl;
        file.close();
    }
}

// Изменение статуса заказа
void Order::ChangeStatus(const std::string& newStatus) {
    statusOrder = newStatus;
    if (newStatus == "Оплачен") {
        cout << "\n=== ВАШ ЗАКАЗ УСПЕШНО ОФОРМЛЕН! ===" << endl;
        cout << "Номер заказа: " << numberOrder << endl;
        cout << "Статус: " << statusOrder << endl;
        cout << "==================================" << endl;
    }
    else {
        cout << "Статус заказа изменен на: " << statusOrder << endl;
    }
}

// Оплата заказа
bool Order::Payment() {
    cout << "\n=== ОПЛАТА ЗАКАЗА ===" << endl;
    cout << "Сумма к оплате: " << costOrder << " руб." << endl;

    int paymentChoice;
    bool validChoice = false;

    while (!validChoice) {
        cout << "Выберите способ оплаты:" << endl;
        cout << "1. Банковская карта" << endl;
        cout << "2. Электронный кошелек" << endl;
        cout << "3. Наложенный платеж" << endl;
        cout << "Ваш выбор: ";
        cin >> paymentChoice;

        if (cin.fail()) {
            cin.clear();
            cin.ignore(numeric_limits<streamsize>::max(), '\n');
            cout << "Ошибка! Введите число от 1 до 3." << endl;
            continue;
        }

        if (paymentChoice >= 1 && paymentChoice <= 3) {
            validChoice = true;
        }
        else {
            cout << "Неверный выбор! Введите 1, 2 или 3." << endl;
        }
    }

    switch (paymentChoice) {
    case 1:
        payWay = "Банковская карта";
        break;
    case 2:
        payWay = "Электронный кошелек";
        break;
    case 3:
        payWay = "Наложенный платеж";
        break;
    }

    cout << "Обработка платежа..." << endl;
    cout << "Способ оплаты: " << payWay << endl;
    cout << "Оплата прошла успешно!" << endl;
    return true;
}

// Вывод информации о заказе
void Order::InfoOrder() const {
    cout << "\n=== ИНФОРМАЦИЯ О ЗАКАЗЕ ===" << endl;
    cout << "Номер заказа: " << numberOrder << endl;
    cout << "Клиент: " << clientOrder << endl;
    cout << "Дата заказа: " << dateOrder << endl;
    cout << "Статус: " << statusOrder << endl;
    cout << "Способ оплаты: " << payWay << endl;
    cout << "Количество товаров: " << countOrder << endl;
    cout << "Общая стоимость: " << costOrder << " руб." << endl;
    cout << "Магазин: " << shopOrder << endl;

    if (itemCount > 0) {
        cout << "\nСостав заказа:" << endl;
        cout << "------------------------------------" << endl;
        for (int i = 0; i < itemCount; i++) {
            if (orderItems[i] != nullptr) {
                cout << (i + 1) << ". " << orderItems[i]->getName()
                    << " - " << quantities[i] << " шт. x "
                    << orderItems[i]->getPrice() << " руб. = "
                    << (quantities[i] * orderItems[i]->getPrice()) << " руб." << endl;
            }
        }
        cout << "------------------------------------" << endl;
    }
    cout << "==================================" << endl;
}

// Генерация номера заказа
std::string Order::generateOrderNumber() {
    srand(static_cast<unsigned int>(time(0)));
    int randomNum = rand() % 10000;
    return "ORD" + std::to_string(randomNum);
}

// Получение текущей даты
std::string Order::getCurrentDate() {
    time_t now = time(0);

#ifdef _MSC_VER
    struct tm timeinfo;
    localtime_s(&timeinfo, &now);
    char buffer[80];
    strftime(buffer, 80, "%d.%m.%Y %H:%M:%S", &timeinfo);
#else
    tm* localTime = localtime(&now);
    char buffer[80];
    strftime(buffer, 80, "%d.%m.%Y %H:%M:%S", localTime);
#endif

    return std::string(buffer);
}

File: 6lab/cs/Order.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;

namespace MusicStore
{
    public class Order
    {
        private const int MAX_ORDER_ITEMS = 100;
        private string numberOrder;
        private string clientOrder;
        private Product[] orderItems;
        private int[] quantities;
        private int itemCount;
        private int countOrder;
        private double costOrder;
        private string statusOrder;
        private string payWay;
        private string shopOrder;
        private string dateOrder;

        // Конструктор по умолчанию
        public Order()
        {
            orderItems = new Product[MAX_ORDER_ITEMS];
            quantities = new int[MAX_ORDER_ITEMS];
            itemCount = 0;
            countOrder = 0;
            costOrder = 0.0;
            numberOrder = GenerateOrderNumber();
            statusOrder = "Ожидает оплаты";
            dateOrder = GetCurrentDate();
            shopOrder = "Музыкальный магазин 'MusicLoverzz'";
            payWay = "Не выбран";

            // Инициализация массивов
            for (int i = 0; i < MAX_ORDER_ITEMS; i++)
            {
                orderItems[i] = null;
                quantities[i] = 0;
            }

            Console.WriteLine($"Вызван конструктор по умолчанию Order: {numberOrder}");
        }

        // Конструктор с параметрами
        public Order(string clientId, Basket basket)
        {
            orderItems = new Product[MAX_ORDER_ITEMS];
            quantities = new int[MAX_ORDER_ITEMS];
            itemCount = 0;
            countOrder = 0;
            costOrder = 0.0;
            clientOrder = clientId;
            numberOrder = GenerateOrderNumber();
            statusOrder = "Ожидает оплаты";
            dateOrder = GetCurrentDate();
            shopOrder = "Музыкальный магазин 'MusicLoverzz'";
            payWay = "Не выбран";

            // Инициализация массивов
            for (int i = 0; i < MAX_ORDER_ITEMS; i++)
            {
                orderItems[i] = null;
                quantities[i] = 0;
            }

            CopyProductsFromBasket(basket);
            Console.WriteLine($"Вызван конструктор с параметрами Order: {numberOrder}");
        }

        // ЗАПРЕЩЕННЫЙ конструктор копирования
        private Order(Order other)
        {
           
            throw new InvalidOperationException("Копирование Order запрещено!");
        }

        
        ~Order()
        {
            Console.WriteLine($"Вызван деструктор Order: {numberOrder}");
            ClearOrderItems();
        }

        // Методы
        public void ChangeStatus(string newStatus)
        {
            statusOrder = newStatus;
            if (newStatus == "Оплачен")
            {
                Console.WriteLine("\n=== ВАШ ЗАКАЗ УСПЕШНО ОФОРМЛЕН! ===");
                Console.WriteLine($"Номер заказа: {numberOrder}");
                Console.WriteLine($"Статус: {statusOrder}");
                Console.WriteLine("==================================");
            }
            else
            {
                Console.WriteLine($"Статус заказа изменен на: {statusOrder}");
            }
        }

        public bool Payment()
        {
            Console.WriteLine("\n=== ОПЛАТА ЗАКАЗА ===");
            Console.WriteLine($"Сумма к оплате: {costOrder:F2} руб.");

            int paymentChoice;
            bool validChoice = false;

            while (!validChoice)
            {
                Console.WriteLine("Выберите способ оплаты:");
                Console.WriteLine("1. Банковская карта");
                Console.WriteLine("2. Электронный кошелек");
                Console.WriteLine("3. Наложенный платеж");
                Console.Write("Ваш выбор: ");

                if (!int.TryParse(Console.ReadLine(), out paymentChoice))
                {
                    Console.WriteLine("Ошибка! Введите число от 1 до 3.");
                    continue;
                }

                if (paymentChoice >= 1 && paymentChoice <= 3)
                {
                    validChoice = true;
                }
                else
                {
                    Console.WriteLine("Неверный выбор! Введите 1, 2 или 3.");
                }
            }
            return true;
        }
        public void InfoOrder()
        {
            Console.WriteLine("\n=== ИНФОРМАЦИЯ О ЗАКАЗЕ ===");
            Console.WriteLine($"Номер заказа: {numberOrder}");
            Console.WriteLine($"Клиент: {clientOrder}");
            Console.WriteLine($"Дата заказа: {dateOrder}");
            Console.WriteLine($"Статус: {statusOrder}");
            Console.WriteLine($"Способ оплаты: {payWay}");
            Console.WriteLine($"Количество товаров: {countOrder}");
            Console.WriteLine($"Общая стоимость: {costOrder:F2} руб.");
            Console.WriteLine($"Магазин: {shopOrder}");

            if (itemCount > 0)
            {
                Console.WriteLine("\nСостав заказа:");
                Console.WriteLine("-----------------------------------------------");
                for (int i = 0; i < itemCount; i++)
                {
                    if (orderItems[i] != null)
                    {
                        Console.WriteLine($"{i + 1}. {orderItems[i].GetName()} - " +
                                          $"{quantities[i]} шт. x " +
                                          $"{orderItems[i].GetPrice():F2} руб. = " +
                                          $"{(quantities[i] * orderItems[i].GetPrice()):F2} руб.");
                    }
                }
                Console.WriteLine("-----------------------------------------------");
            }
            Console.WriteLine("==================================");
        }

        public static bool MakeOrder(string clientId, Basket basket)
        {
            if (basket.IsEmpty())
            {
                Console.WriteLine("Корзина пуста! Добавьте товары перед оформлением заказа.");
                return false;
            }

            Console.WriteLine("\n=== ПОДТВЕРЖДЕНИЕ ЗАКАЗА ===");
            basket.DisplayBasket();

            Console.Write("\nВы точно хотите оформить заказ? (y/n): ");
            char confirmation = Console.ReadKey().KeyChar;
            Console.WriteLine();

            if (confirmation != 'y' && confirmation != 'Y')
            {
                Console.WriteLine("Оформление заказа отменено.");
                return false;
            }

            // Создаем заказ
            Order newOrder = new Order(clientId, basket);

            if (newOrder.Payment())
            {
                newOrder.ChangeStatus("Оплачен");
                Console.WriteLine("\n=== ЗАКАЗ УСПЕШНО ОФОРМЛЕН ===");
                newOrder.InfoOrder();
                newOrder.SaveToFile();
                basket.ClearBasket();
                return true;
            }
            else
            {
                newOrder.ChangeStatus("Ошибка оплаты");
                Console.WriteLine("Ошибка при оформлении заказа!");
                return false;
            }
        }

        public void SaveToFile()
        {
            try
            {
                using (StreamWriter file = File.AppendText("orders.txt"))
                {
                    file.Write($"{numberOrder}|{clientOrder}|{statusOrder}|{costOrder:F2}");

                    for (int i = 0; i < itemCount; i++)
                    {
                        if (orderItems[i] != null)
                        {
                            file.Write($"|{orderItems[i].GetArticle()}|{quantities[i]}");
                        }
                    }
                    file.WriteLine();
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Ошибка сохранения заказа: {ex.Message}");
            }
        }

        // Геттеры
        public string NumberOrder => numberOrder;
        public string ClientOrder => clientOrder;
        public string StatusOrder => statusOrder;
        public double CostOrder => costOrder;
        public int CountOrder => countOrder;
        public string DateOrder => dateOrder;
        public string PayWay => payWay;
        public string ShopOrder => shopOrder;
        public int ItemCount => itemCount;

        public Product GetOrderItem(int index)
        {
            if (index >= 0 && index < itemCount)
            {
                return orderItems[index];
            }
            return null;
        }

        public int GetQuantity(int index)
        {
            if (index >= 0 && index < itemCount)
            {
                return quantities[index];
            }
            return 0;
        }

        // Приватные методы
        private void CopyProductsFromBasket(Basket basket)
        {
            if (basket.IsEmpty())
            {
                return;
            }

            itemCount = basket.ProductCount;
            countOrder = (int)basket.GetBasketCount(); // Исправлено: приведение типа
            costOrder = basket.GetBasketCost();

            for (int i = 0; i < itemCount; i++)
            {
                Product basketProduct = basket.GetProduct(i);
                if (basketProduct != null)
                {
                    // Создаем новый объект Product с теми же данными
                    orderItems[i] = new Product(
                        basketProduct.GetArticle(),
                        basketProduct.GetName(),
                        basketProduct.GetBrand(),
                        basketProduct.GetProducerCountry(),
                        basketProduct.GetCategory(),
                        basketProduct.GetTypeProduct(),
                        basketProduct.GetDescription(),
                        basketProduct.GetQuantityStock(),
                        basketProduct.GetPrice()
                    );
                    quantities[i] = basket.GetQuantity(i);
                }
            }
        }

        private void ClearOrderItems()
        {
            for (int i = 0; i < itemCount; i++)
            {
                orderItems[i] = null;
                quantities[i] = 0;
            }
            itemCount = 0;
        }

        private string GenerateOrderNumber()
        {
            Random rand = new Random();
            int randomNum = rand.Next(10000);
            return $"ORD{randomNum:D4}";
        }

        private string GetCurrentDate()
        {
            return DateTime.Now.ToString("dd.MM.yyyy HH:mm:ss");
        }
    }
}